Version Control section populated

PeterTurcan opened the pull request:
@anarthal Not sure I have included everything you noted for version control - I could not find any generic method for finding a particular version of a library - release notes or the meta information was not consistent - but perhaps it is not needed if the Boost version is known. Let me know if more content is needed. Do individual contributors need to know more about the tools for the Super-project perhaps?
@alandefreitas  fyi

alandefreitas commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+
+The *master* branch of the Super-project is usually the development version of Boost, where new features are added and bugs are fixed. When there is a public release of Boost, it is built from the Super-project repo.
+
+== Bug and Issue Reporting
```
Isn't "Bug and Issue Reporting" more related to the user guide?

anarthal replied:
Yeah, this extra info should go under "Reporting Issues" in User guide.

I think there's room for a section like this here, but geared towards library authors. Probably speaking on how to create an issue template in GitHub to guide users and make the process smoother. If you prefer, I can take the lead on this, apply it to MySQL and submit a PR with whatever I find better.

alandefreitas commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+
+When reporting a bug or issue, you might be offered a range of templates to choose from to guide your filing. If there is no template to follow, cut-and-paste the following into the text field for the issue, and replace the text in italics with appropriate details.
+
+### Version of <Library/Boost>
```
This can go into a markdown code block.

alandefreitas commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+
+The more information you provide the sooner your issue can get resolved.
+
+Note:: Currently, Boost does not use the Git Discussion feature. If filing an issue does not seem appropriate, post on the https://lists.boost.org/mailman/listinfo.cgi/boost[Boost developers mailing list].
```
Discussions is a GitHub feature. Not git.

PeterTurcan commented:
@alandefreitas good feedback, thanks - I will work your comments into the topic

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
 :navtitle: Version Control
 
-*Content under construction.*
+Boost uses the Git version control system for its development. This allows multiple contributors to work on the libraries concurrently, while keeping track of the changes that each person makes. Git is a distributed version control system, which means each contributor has their own local copy of the entire project, including its history.
```
I don't know if placing this under "Releases" is the best choice. It took me some time to find it. I don't know if it fits in any of the current categories, but feels weird under releases, since you use it for regular development.

PeterTurcan replied:
@anarthal Yes, perhaps a new menu entry "Development" would work, adding this and other programming/development topics?

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+
+The Boost libraries are modular, and each library has its own separate repository.  This makes it easier to work with the entire collection of Boost libraries.
+
+The official Boost repositories are hosted on GitHub at https://github.com/boostorg/[Boost.org]. Each library has its own repository under this organization. For example, https://github.com/boostorg/json for boost:json[], or https://github.com/boostorg/date_time/ for boost:date_time[].
```
I'd phrase this as "... are hosted in GitHub under the https://github.com/[boostorg] account". As it's phrased, it looks like we have our own GitHub server hosted under boost.org.
The link https://github.com/boostorg/[Boost.org] seems to redirect to the `boostorg` account URL I've listed above, so better use that.

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+
+The official Boost repositories are hosted on GitHub at https://github.com/boostorg/[Boost.org]. Each library has its own repository under this organization. For example, https://github.com/boostorg/json for boost:json[], or https://github.com/boostorg/date_time/ for boost:date_time[].
+
+As with any Git-based project, you can clone the repositories, make changes, commit those changes to your local repository, and push your changes back to the server. If you want to contribute changes to the official Boost libraries, you can make a pull request, which the maintainers of the libraries can then review and potentially merge into the official codebase. Be aware though that Boost has a ...
```
Maybe link to the "Contribute to an Existing Library" subsection of getting started instead?

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+
+The Boost https://github.com/boostorg/boost[*Super-project*] repository includes all the libraries as submodules.
+
+In addition to the libraries, you'll find many meta-repositories under `boostorg` that handle other tasks, for example:
```
The term meta-repositories suggests that they are not real repositories. I'd phrase it as ."...you'll find many other repositories..."

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+* https://github.com/boostorg/regression[Regression testing]
+* https://github.com/boostorg/release-tools[Release tools]
+
+There is also a https://github.com/boostorg/boost/tree/master/tools[tools] directory in the Super-project, which contains various tools used for tasks such as building and testing Boost libraries. 
```
This is slightly misleading, since some of the projects you listed above (e.g. `cmake`) are actually submodules located under `tools`.

PeterTurcan replied:
@anarthal OK, will rework this

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+
+There is also a https://github.com/boostorg/boost/tree/master/tools[tools] directory in the Super-project, which contains various tools used for tasks such as building and testing Boost libraries. 
+
+The *master* branch of the Super-project is usually the development version of Boost, where new features are added and bugs are fixed. When there is a public release of Boost, it is built from the Super-project repo.
```
I'd suggest to rework & extend this a little bit.

* Each individual repository has a development (`develop`) branch, and a "this will go to prod the next release" branch (`master`). I'd say all the repos follow this, but you'd have to confirm it in Slack.
* The superproject does have these two branches: develop and master, with the same semantics.
* (git background) In git, when you add a submodule to a project, you make it point to a particular commit. This means that, when devs update sub-projects, the superproject doesn't get immediately updated.
* For this reason, there is a commit bot that runs nightly and updates the commit to which each submodule in the superproject points to. IIRC, this happens nightly and for both the `develop` and `master` branches. Please check the details on Slack too, since I do not maintain this.
* What are the implications of merging something in your library/submodule to develop? Due to how Boost is tested, you may affect other devs. As part of Boost testing, CIs usually clone the superproject and any required submodules. When testing any branch that is not master, the `develop` branch of the superproject is cloned. When you're testing `master` or a release, the `master` branch of the superproject is cloned. While this is not a requirement, most libs do it.
* Following the argument above, merging something to `develop` may affect any dependent code in the rest of Boost as soon as the commit bot makes the update to the superproject. E.g. if I do something wrong to MySQL develop branch, I will affect the `cmake` module, which runs tests to verify that you can use MySQL via cmake.
* In addition to what's listed above, the `master` branch dictates what will be in the next release. This is the branch that is frozen during the release process.
* It may be noteworthy that you, as a dev, don't directly create tags under repositories, ever. This differs from the usual workflow in other projects. Tags are created by the release scripts, both in the superproject and the individual repos.
* If you need to look at the code for the `1.82.0` release, just `git checkout boost-1.82.0` in the relevant repo.

PeterTurcan replied:
@anarthal - good info

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+
+== Bug and Issue Reporting
+
+Apart from the code repositories, the `boostorg` organization also hosts other resources related to Boost development, such as issue trackers for each library, and discussion forums (in the form of GitHub Discussions).
```
Issue trackers and discussions are part of GitHub. So I'd say "repositories use the built-in GitHub issue tracker and discussions facilities".

Note too that not all repos use the discussions thing. Vinnie doesn't like it, for instance. I try my users to discuss stuff via issues, since it centralizes things.

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,57 @@
+
+### Version of <Library/Boost>
+
+_If you can find the version of the library itself (in a readme file, other documentation, or a header file), then put that version number here. Otherwise, if you have difficulty finding a version for the library itself, add the Boost version number, such as `1.82.0`._
```
If you gear this towards a user, I'd tell them to just include the Boost version number, since I'd say the author knows the internal version (if any) for a given Boost version.

PeterTurcan replied:
@anarthal Good point on the issue templates. I have the basics of how to create one, but it would be great to point out a good example - so please go ahead and add one to the MySQL, and I'll link to it. Thanks

PeterTurcan commented:
@alandefreitas @anarthal  Addressed your feedback, notably:
1 Moved the issue filling guidance to the user guide Reporting Issues section
2 Added a section on creating an Issue Template
3 Added a section on Breaking Changes
4 Added cloning practices to the super project section

alandefreitas commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -7,51 +7,125 @@ Boost uses the Git version control system for its development. This allows multi
+CMake is one of the most popular of the https://github.com/boostorg/boost/tree/master/tools[boost/tools] directory. You will find other tools used for tasks such as building and testing Boost libraries. 
 
-The *master* branch of the Super-project is usually the development version of Boost, where new features are added and bugs are fixed. When there is a public release of Boost, it is built from the Super-project repo.
+The Super-project has both *master* and *develop* branches, which have the same semantics as these branches do for individual libraries. New features are added and bugs are fixed in the *develop* branch. When the *develop* branch is complete, it is merged with the *master* branch. When there is a public release of Boost, it is built from the *master* branch after that branch has been closed to up ...
```
> When the *develop* branch is complete, it is merged with the *master* branch.

This is unfortunately technically incorrect.

- For individual libraries, when the *develop* branch is complete, it is merged with the *master* branch.
- For the super-project, the develop and master branches are independent. They track the latest changes in master/develop in all libraries. There's no merging involved because this strategy makes merging impossible.

I argued against this strategy myself but it's how it is.

PeterTurcan replied:
Ah, OK, added an explanation

alandefreitas commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -7,51 +7,125 @@ Boost uses the Git version control system for its development. This allows multi
+=== Working with the Super-project
 
-Apart from the code repositories, the `boostorg` organization also hosts other resources related to Boost development, such as issue trackers for each library, and discussion forums (in the form of GitHub Discussions).
+It is good practice to create submodules to a project, when developing your library.
```
Is it?

PeterTurcan replied:
Not sure, so reworded

alandefreitas commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -7,51 +7,125 @@ Boost uses the Git version control system for its development. This allows multi
+Boost maintains its release versions with https://github.com/boostorg/boost/tags[Tags], where the tag is always the `boost-` prefix followed by a https://semver.org/[Semantic Versioning] string.  Using semver strings is useful because that's something most tools understand, and for most users who just want to find a certain boost version, that's enough information.
 
-The more information you provide the sooner your issue can get resolved.
+But for users who want to know if something broke from one version to another, that's where things get complicated. The semver specification  requires an increase in the major version number (for example, from `1.2.3` to `2.0.0`) if something breaks the API. The minor version should be increased (for example: `1.81.0` to `1.82.0`) if there are new features.
```
Expressions like "to know if something broke" looked OK for a quick message but are a little too informal to be in the documentation.

PeterTurcan replied:
agreed, reworded

alandefreitas commented on user-guide/modules/ROOT/pages/reporting-issues.adoc:
```diff
@@ -9,14 +9,48 @@ If you come across a bug, or other issue, with one of the libraries, go through
+
+The more information you provide the sooner your issue can get resolved.
+
+Note:: Currently, Boost does not use the GitHub Discussion feature. If filing an issue does not seem appropriate, post on the https://lists.boost.org/mailman/listinfo.cgi/boost[Boost developers mailing list].
```
-> The boost super-project repository does not use the GitHub Discussion feature

PeterTurcan replied:
ok, updated

PeterTurcan commented:
@alandefreitas thanks for the additional feedback - see if my updated explanations work?

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,138 @@
+
+=== Working with the Super-project
+
+You might consider creating submodules for a project, when developing your library. This is not a required process, but it does fit with some of the Super-project tooling:
```
What does this mean? Submodules are usually only created in the super-project, and regular developers don't usually do it directly, but some specific members of Boost do.

PeterTurcan replied:
ah, got it

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,138 @@
+
+* https://github.com/boostorg/website2022[Website generation]
+* https://github.com/boostorg/cmake[CMake support]
+* https://github.com/boostorg/regression[Regression testing]
```
Reviewing this, I think `tools/build` (which holds the B2 sources) may be more important than website (that will probably disappear soon) or regression (which I think it's extremely old)

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,138 @@
+* https://github.com/boostorg/regression[Regression testing]
+* https://github.com/boostorg/release-tools[Release tools]
+
+CMake is one of the most popular of the https://github.com/boostorg/boost/tree/master/tools[boost/tools] directory. You will find other tools used for tasks such as building and testing Boost libraries. 
```
"the most popular directory" sounds a little bit weird - is this entire paragraph really required?

PeterTurcan replied:
removed the para

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,138 @@
+
+The Super-project has both *master* and *develop* branches, which operate somewhat differently than for individual libraries. 
+
+New features are added and bugs are fixed in library *develop* branches. When the *develop* branch is complete, it is merged with the library *master* branch. 
```
A branch doesn't get "complete" - code changes in develop are tested and once the code is considered to be stable (e.g. CIs have completed, other projects are building without errors, or whatever your criteria is), then it's merged to master.

PeterTurcan replied:
ok, reworded

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,138 @@
+
+You might consider creating submodules for a project, when developing your library. This is not a required process, but it does fit with some of the Super-project tooling:
+
+* In Git, when you add a submodule to a project, you make it point to a particular commit. This means that, when developers update their sub-projects, the Super-project _doesn't_ get immediately updated. For this reason, there is a commit bot that runs nightly and updates the commit to which each submodule references in the Super-project. This happens nightly and for both the *develop* and *maste ...
```
I think there's a problem with ordering here. 

> For the Super-project, the *develop* and *master* branches are independent. Both branches track the latest changes in the corresponding library branch, for all libraries. The Super-project branches are never merged, as this strategy prevents merging. In other words, the Super-project *master* branch is created from the library *master* branches, not from the Super-project *develop* branch (which is primarily used for testing).

This is a direct consequence of the paragraph in this comment, but it appears before the cause.

I'd re-phrase the git comment as "recall that..." since I'd expect the audience of this page to already know how submodules work (although a refreshing may be good).

PeterTurcan replied:
ok

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,138 @@
+
+* In Git, when you add a submodule to a project, you make it point to a particular commit. This means that, when developers update their sub-projects, the Super-project _doesn't_ get immediately updated. For this reason, there is a commit bot that runs nightly and updates the commit to which each submodule references in the Super-project. This happens nightly and for both the *develop* and *maste ...
+
+* The implications of merging something in your library or submodule to the Super-project *develop* branch, include that due to how Boost is tested, you may affect other developers. As part of Boost testing, Continuous Integration (CI) usually clones the Super-project and any required submodules. 
```
You don't merge to the super-project develop branch. You merge to your project's develop branch, and the commit bot will update the super-project.

PeterTurcan replied:
OK, got it

anarthal commented on contributor-guide/modules/ROOT/pages/version-control.adoc:
```diff
@@ -1,4 +1,138 @@
+
+== Breaking Changes
+
+Boost maintains its release versions with https://github.com/boostorg/boost/tags[Tags], where the tag is always the `boost-` prefix followed by a https://semver.org/[Semantic Versioning] string.  Using semver strings is useful because that's something most tools understand, and for most users who just want to find a certain boost version, that's enough information.
```
I think we've got a little bit of a contradiction in this section - we're essentially saying "we use semantic versioning, except that we don't use semantic versioning at all"

PeterTurcan replied:
Agreed, reworded.

PeterTurcan commented:
@anarthal @alandefreitas Your feedback included, and related CI section added

anarthal commented:
@alandefreitas @PeterTurcan the version control looks good. The CI section will likely require a couple changes or additions (since it's new) but better do that in a different branch/story.

PeterTurcan commented:
@alandefreitas squashed!
